### PR TITLE
Fixes Brazilian Daylights Savings Time duplication bug in Firefox

### PR DESCRIPTION
**Pull request by majapw**
Fix for https://github.com/airbnb/react-dates/issues/55

@ljharb @wmartins 

Also, for a fun time, take a look at https://en.wikipedia.org/wiki/Daylight_saving_time_by_country

Did you know that Greenland does DST at 10PM in the fall and 11PM in the spring? Did you know that not all of Brazil uses Daylight Savings Time? Did you know that Arizona does not have Daylight Savings Time, even though the rest of the USA does, but that the Navajo Nation Community inside of Arizona does??? U.U

### CODE REVIEW

File: src/utils/getCalendarMonthWeeks.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 export default function getCalendarMonthWeeks(month, enableOutsideDays) {
-  const firstOfMonth = month.clone().startOf('month');
-  const lastOfMonth = month.clone().endOf('month');
+  const firstOfMonth = month.clone().startOf('month').add(2, 'hours').add(1, 'minute');
```
Comment (ljharb) [inline src/utils/getCalendarMonthWeeks.js]: it would be great to add a test with the brazil moment locale :-)

**Comment by ljharb** [comment]
LGTM

**Comment by wmartins** [comment]
Hey @majapw, the code fixes it, but I don't know if it is the best solution for that.

For example, if you set your timezone to "Europe/Athens" you end up having things like that:



(Displaying 1 extra day, and with external days prop  as false)

**Comment by ljharb** [comment]
hmm, there's probably other logic that needs adjusting to take the extra 121 minutes into account.

**Comment by majapw** [comment]
Yeah. Good catch! I will dig around more into this. :)

**Comment by wmartins** [comment]
@ljharb @majapw use utc in this case is an option?

Setting `firstOfMonth` and `lastOfMonth` to utc fixes it. The downside is that the following tests will break:

```
1) getCalendarMonthWeeks enableOutsideDays arg is false first non-null element is first of the month
2) getCalendarMonthWeeks enableOutsideDays arg is false last non-null element is last of the month
3) getCalendarMonthWeeks enableOutsideDays arg is true contains last of the month
```

I will check it out if this is something specific to tests or if it is really breaking things.

**Comment by majapw** [comment]
Those tests should only be comparing by day though... that seems concerning.

**Comment by wmartins** [comment]
@majapw, sorry, totally my bad, I was just testing that using your branch (I modified your modified code), let me try that in `master` branch

**Comment by wmartins** [comment]
@majapw, tested the `utc` solution on `master` (now I did it correctly) and everything works fine (with `America/Sao_Paulo` and `Europe/Athens` TZs). Also, the tests are passing. :)

**Comment by majapw** [comment]
So I changed `getCalendarMonthWeeks` to use utc locally as you suggested @wmartins, but my tests don't seem to have been affected.

**Comment by wmartins** [comment]
I'll run more tests here to see if there is nothing wrong

**Comment by wmartins** [comment]
@majapw, I tested here and it is showing a an extra day in month, I think you need to change the order of `utc` calling. Because your code is setting it to utc after getting the start of month.

Change those:

```javascript
const firstOfMonth = month.clone().startOf('month').utc();
const lastOfMonth = month.clone().endOf('month').utc();
```

to:

```javascript
const firstOfMonth = month.clone().utc().startOf('month');
const lastOfMonth = month.clone().utc().endOf('month');
```

Here is the problem occurring:

**Comment by majapw** [comment]
I think the edge case I'm concerned about is... let's say I am in PT, and it is 5:01PM on August 31st.

`moment()` returns Aug 31st. Does `moment.utc()` return September 1st? If so, would calling `getCalendarMonthWeeks` at that time actually return the weeks for September?

**Comment by wmartins** [comment]
Yeah, this could be a problem, depending on how this method is being
called. We can try to "build" a new "base date" (an utc one) based on the
month (the current user date, non utc). And then get the start of month and
end of month. It will guarantee that the month is correct.

What do you think about it?

Em sex, 2 de set de 2016 20:08, Maja Wichrowska <notifications@github.com>
escreveu:

**Comment by majapw** [comment]
Yeah, I was thinking that the best bet is probably to extract the month information from the passed in object and then construct an entirely new moment object.... and then apply the utc method to that. Yuck!

**Comment by wmartins** [comment]
Yep. Another option is to check if moment has some "set to utc without
modifying the date" method. If so, it would preserve the date, then it
would be safe to use it.

Em sex, 2 de set de 2016 20:44, Maja Wichrowska <notifications@github.com>
escreveu:

**Comment by wmartins** [comment]
Hey @majapw, I think I've found a better solution for this!

First, the problem is occurring because we are building today (under GMT-03:00) all the dates in the "future", but, `moment#add` is messing things up when the date is not an utc date. So, it adds one more day when the timezone is changed through the year (in this case, at October 15th). What we can do to mitigate this is to use utc dates in `firstOfMonth` and `lastOfMonth`, but then we would end up with the problems talked before (changing the month based on the current date and screwing up everything).

I have a better solution that doesn't screw up everything:

1. Create a `baseDate`, that is a clone of the `month` object, but with the [`utcOffset`](http://momentjs.com/docs/#/manipulating/utc-offset/) **fixed**
2. Now, when calling `moment#add`, moment knows how to do the calculation (https://github.com/moment/moment/issues/2353#issuecomment-97470780). Using `utcOffset` will "flag" this date as utc

Here is the code I used:

```javascript

const baseDate = month.clone().utcOffset(month.utcOffset());
const firstOfMonth = baseDate.clone().startOf('month');
const lastOfMonth = baseDate.clone().endOf('month');
```

It seems more reliable, and it is not that hard to understand, because you're basically saying "ok moment, use that timezone, do all the math you need and display things based on that timezone".

What do you think?

**Comment by majapw** [comment]
This sounds amazing! If you want to put up a PR for that (maybe with a comment explaining the reasoning), I would be happy to close this one and merge that in. :)

**Comment by wmartins** [comment]
Oh, it would be a pleasure, here is the PR @majapw: #58!

**Pull request closed without merging**